components/lib/jexl/src/main/java/mozilla/components/lib/jexl/evaluator/EvaluatorHandlers.kt

```diff
@@ -0,0 +1,145 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.lib.jexl.evaluator
+
+import mozilla.components.lib.jexl.ast.AstNode
+import mozilla.components.lib.jexl.ast.AstType
+import mozilla.components.lib.jexl.value.JexlArray
+import mozilla.components.lib.jexl.value.JexlBoolean
+import mozilla.components.lib.jexl.value.JexlDouble
+import mozilla.components.lib.jexl.value.JexlInteger
+import mozilla.components.lib.jexl.value.JexlObject
+import mozilla.components.lib.jexl.value.JexlString
+import mozilla.components.lib.jexl.value.JexlUndefined
+import mozilla.components.lib.jexl.value.JexlValue
+
+internal typealias EvaluatorFunction = (Evaluator, AstNode) -> JexlValue
+
+/**
+ * A mapping of AST node type to a function that can evaluate this type of node.
+ *
+ * This mapping could be moved inside [Evaluator].
+ */
+internal object EvaluatorHandlers {
+    private val functions: Map<AstType, EvaluatorFunction> = mapOf(
+        AstType.LITERAL to { _, node ->
+            val value = node.value
+            when (value) {
+                is String -> JexlString(value)
+                is Double -> JexlDouble(value)
+                is Int -> JexlInteger(value)
+                is Boolean -> JexlBoolean(value)
+                else -> throw EvaluatorException("Unknown value type: ${value!!::class}")
+            }
+        },
+
+        AstType.BINARY_EXPRESSION to { evaluator, node ->
+            val left = evaluator.evaluate(node.left!!)
+            val right = evaluator.evaluate(node.right!!)
+
+            val operator = evaluator.grammar.elements[node.operator!!]!!
```

Comment:
We could fix this `[node.operator!!]!!` when we refactor `AstNode` into different types. Then can we verify that node is a BinaryExpressionNode where `operator` is not nullable.

Reply:
Yeah, that's my biggest complaint about the current implementation too. I would like to have subtypes for `AstNode`. Some properties may still need to be nullable though as they get filled while parsing and are not available yet at the point of creating the node.

However that's a very big refactoring and I would like to do that after landing the first version. :)